### PR TITLE
CLI Fork Exec

### PR DESCRIPTION
**Pull request by gronke**
- Exposes Jail.fork_exec() via the CLI.
- Attach arbitrary data to IocageEvent on message update

### CODE REVIEW

File: daemon.py [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+import http.server
+import urllib.parse
+
+jail = iocage.Jail('foo')
+
+class ForkExecDaemon(http.server.BaseHTTPRequestHandler):
+
+    def do_GET(self):
+        
+        command = urllib.parse.unquote(self.path.strip("/")).strip(" ").split(" ")
```
Comment (gronke) [inline daemon.py]: Bandit and @igalic will love this!

File: iocage/lib/events.py [hunk 2]
```diff
@@ -176,8 +176,11 @@ def duration(self) -> typing.Optional[float]:
             return None
 
     def _update_message(self, **kwargs) -> None:
-        if "message" in kwargs:
-            self.message = kwargs["message"]
+        for key in list(kwargs):
```
Comment (skarekrow) [inline iocage/lib/events.py]: This could be `for key in kwargs.keys:` no?

**Comment by gronke** [comment]
### On the server
```
$ ioc create -b -n fork-exec-test exec_start="/usr/bin/true"               
fork-exec-test successfully created from 11.1-RELEASE!
$ python3.6 experiments/socket-based-daemon.py fork-exec-test
whoami && hostname
127.0.0.1 - - [01/Jan/1970 13:37:00] "POST / HTTP/1.1" 200 -
```

### On the client
```
$ /usr/bin/time -l curl -i --data "whoami && hostname" 127.0.0.1:8080/
HTTP/1.0 200 OK
Server: BaseHTTP/0.6 Python/3.6.4
Date: Sat, 24 Feb 2018 13:16:14 GMT
Content-type: text/plain

root
fork-exec-test
0.30 real         0.00 user         0.00 sys
      6480  maximum resident set size
       192  average shared memory size
         8  average unshared data size
       128  average unshared stack size
       317  page reclaims
         0  page faults
         0  swaps
         3  block input operations
         0  block output operations
         1  messages sent
         2  messages received
         0  signals received
         6  voluntary context switches
         3  involuntary context switches
```

File: iocage/lib/events.py [hunk 3]
```diff
@@ -176,8 +176,11 @@ def duration(self) -> typing.Optional[float]:
             return None
 
     def _update_message(self, **kwargs) -> None:
-        if "message" in kwargs:
-            self.message = kwargs["message"]
+        for key in kwargs.keys():
+            if key == "message":
+                self.message = kwargs["message"]
+            else:
+                self.data[key] = kwargs[key]
```
Comment (gronke) [inline iocage/lib/events.py]: This allows attaching data to events. In the case of events JailExec we can read stdout and stderr from it when the event has ended.

File: iocage/lib/Jail.py [hunk 4]
```diff
@@ -417,27 +418,41 @@ def _revert_jail_started():
...
         jailExecEvent.add_rollback_step(_revert_jail_started)
 
-        child, stdout, stderr = self.exec(command, ignore_error=True)
-        if child.returncode != 0:
-            if (error_handler is None):
-                passed = False
-                message = ""
+        if passthru is True:
+            try:
+                self.passthru(command)
```
Comment (gronke) [inline iocage/lib/Jail.py]: Passthru enables CLI users to interact with an executed command.

Comment (igalic) [reply iocage/lib/Jail.py]: given how long and unreadable this if/else is, maybe we should also bundle the else's call into a separate function like `passthru`

**Comment by gronke** [comment]
@igalic I've tried refactoring the generative exec approach into another method. Because the steps tightly coupled to the events initialized in the fork_exec method, I think it is not an improvement to break out another private method here.

**Pull request merged**